src/ctap/data_formats_fuzz.rs

```diff
@@ -0,0 +1,98 @@
+use super::{
+    command::AuthenticatorMakeCredentialParameters,
+    data_formats::{
+        CredentialProtectionPolicy, MakeCredentialExtensions, MakeCredentialOptions,
+        PinUvAuthProtocol, PublicKeyCredentialDescriptor, PublicKeyCredentialParameter,
+        PublicKeyCredentialRpEntity, PublicKeyCredentialUserEntity,
+    },
+};
+use arbitrary::{Arbitrary, Unstructured};
+use cbor::{cbor_map, cbor_map_options};
+use sk_cbor as cbor;
+
+impl From<PinUvAuthProtocol> for cbor::Value {
+    fn from(protocol: PinUvAuthProtocol) -> Self {
+        Self::from(protocol as u64)
+    }
+}
+
+pub trait ArbitraryCbor<'a>: Sized {
```

Comment:
ArbitraryCbor is used on types which are not one-to-one corresponding to cbors. For example `MakeCredentialOptions` has two fields but is parsed from a cbor map with three. Also, some fields of the types have defaults in case the map doesn't contain that entry, but some fields are required to exist. If we generate arbitrary types and converting them back to CBOR, we miss some valid CBOR map structures that the type can be parsed from.

Reply:
I see, the structures we use make deserializing from CBOR non-injective. And this trait permits to theoretically generate all CBORs that would successfully deserialize to any possible struct value.

Note though that in practice this is not the case. For example for `MakeCredentialOptions`, the trait implementation always sets all map keys while it would be valid to deserialize when some keys are missing since they have a default value.

Reply:
> Note though that in practice this is not the case. For example for MakeCredentialOptions, the trait implementation always sets all map keys while it would be valid to deserialize when some keys are missing since they have a default value.

What trait implementation? I only see a trait from cbor::Value to MakeCredentialOptions, not reversed. I thought commands from the host are parsed as cbor, and then parsed as MakeCredentialOptions, so we can't guarantee that no host will send a cbor map that has some missing values right?